### DISCUSSION TITLE
How to check whether an image exists before inserting it in the template?

### DISCUSSION THREAD

**Discussion by HeuristicPerson** [getpelican/pelican#3135] [Q&A]
I have a small blog using pelican. For simplicity, I've added a cover image to each of the entries that is called like the main document but with.jpg at the end. Then, from the template, it's easy to show that image.

Now, I have two related problems:

1. I would like to be able to check for the existence of the image, and if it's not there, not insert anything into the final HTML. What would be the best way to achieve that? Modifying the template? Creating a filter? I've tried checking for the existence of the file in the template using os module with something similar to `{% if (os.path.isfile(MY_FILE) %}` but I receive an error that os is not recognised.

2. Similarly, I would like to check for different files and show them in different ways. e.g. having two files called like `entry_title_cover.jpg` and `entry_title_miniature.jpg`; what would be the best solution for that?

My understanding is custom filters (defined or called in `pelicanconf.py`) can only access the data contained within the entries, not in external files like my image paths, so I don't really know how to attack the problem.

Many thanks in advance for your help, and also to the authors of Pelican. I've been using it just for one month or so but I'm really impressed by its simplicity and how weel it works.

**Comment by avaris** [comment]
I would like to see some examples of how you handle these things (how you define them and how template adds them) to give concrete suggestions. For example in (1), I don't get why you would have non-existent files or the need to check the existence of them.

Typical way to handle these is to define a consistent metadata within your content and use that in the templates. i.e. you can provide a `cover: some_file.jpg` in your metadata and in template `{% if content.cover %}{... insert image html ...}{% endif %}` and that would add image if it is defined and skip if not. Similar with `miniature`.

**Reply by justinmayer** [reply]
Indeed, something similar to the latter is what I suggested two days ago: https://github.com/getpelican/pelican/discussions/3133#discussioncomment-5930750

@HeuristicPerson: Glad to hear you are enjoying Pelican. Many thanks for the kind words! 😁

**Reply by HeuristicPerson** [reply]
At the moment, what I do is defining a new variable in the meta section called `footer` which contains the footer of the cover picture I want to show. Then, in `article.html` template I put:

```html
{% if article.footer %}
    <div class="cover">
      <img src="{{ SITEURL }}/images/covers/{{ article.date.strftime('%Y-%m-%d.jpg') }}" />
      <footer>{{ article.footer }}</footer>
    </div>
{% endif %}
```

Ideally, for simplicity, I'd like to include the images without the need of defining extra meta-data tags.

**Reply by justinmayer** [reply]
1. I don't believe there is a non-trivial way to detect whether a file exists in this context.
2. If you always include a cover picture footer, then there is no need for the `{% if article.footer %}` condition. But presumably you already knew that.
3. Where would you store the footer information if not in an extra meta-data tag? Something about what you are trying to accomplish is not-at-all clear.

**Reply by HeuristicPerson** [reply]
The footer meta-tag will store just the text caption of the images, while the image itself is read from the /images/covers folder inside content. This is not robust at all because the user can specify the text caption for a non-existing picture (so in the rendered output you'll see an empty image) or vice-versa (the image is placed in the right place but the user forgot to include the footer meta-data tag, so the image won't appear at all. This is one of the limitations I want to overcome, make the system as simple as possible.

If pelican doesn't support anything like this, I imagine I could write a file pre-processor so I automatically add/remove the tags needed. I'd say it's not ideal in my case but I already have a publishing script that perform some extra actions so that seems the logical place for the new code.

**Reply by avaris** [reply]
I'd say this is still better suited to specific metadata fields. But if you want to automate the association of an image with a content a `plugin` is the way to go. In this context, plugin will do the checking and auto-populate (or omit) these metadata fields. Then template will be as I described.

**Reply by HeuristicPerson** [reply]
Many thanks to all!

I've played a bit with the code this morning and I was able to make the system work passing the "article" variable to the filter and also the context (through a decorator in the filter function). As you said, the system is a bit cumbersome, so I'll try to create a plugin for the task.

Regards.
